### PR TITLE
Make Generator::schedule() optional

### PR DESCRIPTION
**Pull request by steven-johnson**
This officially makes it legal to simply omit the schedule() method from a new-style Generator:

- The advent of late-bound ScheduleParams means that having a separable schedule() method isn't as essential for componentized Generators as it was when this design was first envisioned.
- While a separable schedule() method may prove useful for auto-scheduling, that isn't the case just yet, and (ironically) the presence of this method is serving to hamper conversion of Generators to new-style due to confusion.
- Some existing Generators 'converted' to new-style by putting in an empty schedule() method, putting all the scheduling in the generate() method.

Converted all existing test/generator code to use generate() rather than build() (except for one new test which is explicitly meant to ensure that build()-style still works, as we aren't dropping support for them until all downstream users can be converted).

Note that if a schedule() method is present, it will still be called (so nothing will break from this).

(If this PR is approved, subsequent PRs will convert all remaining Generators in the Halide repo to be new-style, except of course for the one compatibility test.)

### CODE REVIEW

File: apps/conv_layer/conv_layer_generator.cpp [hunk 1]
```diff
@@ -26,8 +28,7 @@ class ConvolutionLayer : public Halide::Generator<ConvolutionLayer> {
 
         f_conv(x, y, z, n) += filter(r.x, r.y, r.z, z) * input(x + r.x, y + r.y, r.z, n);
 
-        Func f_ReLU("ReLU");
-        f_ReLU(x, y, z, n) = max(0, f_conv(x, y, z, n));
```
Comment (abadams) [inline apps/conv_layer/conv_layer_generator.cpp]: "relu" is a meaningful name in this context that we should preserve in some way.

File: apps/cuda_mat_mul/mat_mul_generator.cpp [hunk 2]
```diff
@@ -40,14 +49,9 @@ class MatMul : public Halide::Generator<MatMul> {
             .compute_at(prod, y)
             .vectorize(B.in().args()[0]);
 
-        OutputImageParam bufs[] = {A, B, prod.output_buffer()};
```
Comment (abadams) [inline apps/cuda_mat_mul/mat_mul_generator.cpp]: Is there no suitable base class for input and output? Constrainable?

**Comment by abadams** [comment]
lgtm other than my comments above

File: apps/cuda_mat_mul/mat_mul_generator.cpp (hunk 2 above)
Comment (steven-johnson) [reply apps/cuda_mat_mul/mat_mul_generator.cpp]: Well:
- Constrainable is an abstract class, so we can't use `Constrainable bufs[]`
- arrays-of-reference are illegal in C++, so we can't use `Constrainable &bufs[]`
- We could of course use pointers to do something like
```
        Halide::Internal::Constrainable *bufs[] = {&A, &B, &out};
        for (auto *buf : bufs) {
            buf->set_host_alignment(16)
                .dim(0).set_bounds(0, size)       
                .dim(1).set_stride(size);     
         }     
```
But it's a little non-idiomatic in Halide. (And of course Constrainable is currently Internal but that could be fixed.)

That said, Constrainable could probably be made to be a non-abstract value class with a little bit of reworking, but it's probably better served in a separate PR.

How about I land this as-is, then revisit in a subsequent PR to make Constrainable more useful?

File: apps/conv_layer/conv_layer_generator.cpp (hunk 1 above)
Comment (steven-johnson) [reply apps/conv_layer/conv_layer_generator.cpp]: Fixed.

File: apps/cuda_mat_mul/mat_mul_generator.cpp (hunk 2 above)
Comment (abadams) [reply apps/cuda_mat_mul/mat_mul_generator.cpp]: sgtm

**Pull request merged**